custom_components/hacs/repositories/base.py

```diff
@@ -811,18 +812,18 @@ async def remove_local_directory(self) -> None:
             else:
                 local_path = self.content.path.local
 
-            if os.path.exists(local_path):
+            if await async_exists(self.hacs.hass, local_path):
```

Comment:
One of the things we have sped up the loading in Home Assistant is that we just fetched the folder listing once at the beginning, and check against the variable instead of jumping into the executor to see if a platform exists. We can do that here too, and it will avoid many jumps into the executor.

Reply:
Do you mean we should implement some kind of filesystem cache in hacs to speed up uninstalling of custom components and other rare actions? I *really* don't think that's a good idea.

Reply:
At the top of this function you know that you're going to jump into the executor to test if files exist. You can start with a single executor call to get a file list and avoid all subsequent executor jumps.

Reply:
So what? The function you're commenting on is about removing a custom component, and the specific line is about potentially retrying the remove operation if the path for some reason is locked. It absolutely does not matter if that function makes use of executor threads 4 or 5 times.

Reply:
We should at all times avoid jumps into the executor. It's not just slow code but it uses up slots in the executor that can slow down other integrations. The code we write is copied to custom components, and so we need to ensure its of the highest quality.

If this whole function is for deletion of files on disk, let's run it as a single sync function.

Reply:
This is also my problem with individual async wrappers of file actions, they promote jumping back and forth between async and sync land. If a file exists is often combined with either reading or writing. In that case, it should always combine those two actions in a single executor call.

Reply:
> We should at all times avoid jumps into the executor. It's not just slow code but it uses up slots in the executor that can slow down other integrations
> If this whole function is for deletion of files on disk, let's run it as a single sync function.

I'm sorry, but these statements are contradictory. If we stay in the executor for longer, we use up an executor "slot" for longer.
The line you're commenting on checks if removing the file succeeded, and if it didn't, the code starts a loop of sleep + retry the removal every second.

Also, please note that the description of this PR is:
> Add async wrapper for commonly used filesystem functions `os.path.exists` and `os.remove`

And not:
> Rewrite filesystem accesses to optimize for the number of jumps to executor

Reply:
Refactoring this to be a single jump seems like a good follow-up indeed.
However, for the issue at hand, this solves the problem of calling FS actions from the event loop.

Reply:
Ok